scripts/install-build-tools.sh

```diff
@@ -14,36 +15,210 @@ if (( $EUID != 0 )); then
+            wget https://bootstrap.pypa.io/get-pip.py
+            $SUDO python${PY_VERSION} get-pip.py
+            rm get-pip.py
+        fi
+        # install python3 venv module for linux
+        if ! $SUDO apt install -y "python${PY_VERSION}-venv"; then
+            echo "Failed to install python${PY_VERSION}-venv"
+        else
+            echo "Installed python${PY_VERSION}-venv"
+        fi
+    fi
+    # create virtual environment with specific python version
+    if "${PY_LOC}" -m venv "${ENV_NAME}"; then
+        echo "Virtual environment '${ENV_NAME}' created"
+    else
+        echo "Virtual environment creation failed"; exit 1
+    fi
+    # activate virtual environment
+    source "${ROOT}/scripts/activate-venv.sh"
+    # install python packages
+    if "${ROOT}/${ENV_NAME}/bin/pip" install -r requirements.txt; then
+        echo "Success installing python packages"
+    else
+        echo "Failure installing python packages"
+        deactivate
+        exit 1
+    fi
+    deactivate
+}
+
+
+echo "OS Type: $OSTYPE"
+# macOS install with homebrew
 if [[ "$OSTYPE" == "darwin"* ]]; then
+
+    # macOS does not support running shell scripts as root with homebrew
+    if [ $EUID -eq 0 ]; then
+        echo -e "Mac users should run this script without 'sudo'. Exiting..."
+        exit 1
+    fi
+
     CPUS=$(sysctl -n hw.ncpu)
     # ensure development environment is set correctly for clang
     $SUDO xcode-select -switch /Library/Developer/CommandLineTools
-    brew install llvm@14 googletest google-benchmark lcov make wget cmake curl
+
+    if ! brew --version &>/dev/null; then
+        echo -e "${cyan}Homebrew is required to install dependencies.${end}"
+        exit 1
+    fi
+
+    brew install llvm@14 googletest google-benchmark lcov make wget cmake bash bc
+    brew upgrade bash
+
+    BREW_ROOT=$(brew --prefix)
+
     CLANG_TIDY=/usr/local/bin/clang-tidy
     if [ ! -L "$CLANG_TIDY" ]; then
-        $SUDO ln -s $(brew --prefix)/opt/llvm@14/bin/clang-tidy /usr/local/bin/clang-tidy
+        $SUDO ln -s "${BREW_ROOT}/opt/llvm@14/bin/clang-tidy" /usr/local/bin/clang-tidy
     fi
-
     GMAKE=/usr/local/bin/gmake
     if [ ! -L "$GMAKE" ]; then
         $SUDO ln -s $(xcode-select -p)/usr/bin/gnumake /usr/local/bin/gmake
     fi
-fi
 
-if [[ "$OSTYPE" == "linux-gnu"* ]]; then
-    $SUDO apt update
-    $SUDO apt install -y build-essential wget cmake libgtest-dev libbenchmark-dev lcov git software-properties-common rsync unzip
+    # check if supported version of python3 is already installed, and save the version
+    for PY_VERS in "${PYTHON_VERSIONS[@]}"; do
+        if which "python${PY_VERS}" &> /dev/null; then
+            # save the installed version if found
+            PY_INSTALLED=${PY_VERS}
+            break
+        fi
+    done
+
+    # install valid python version if not installed yet
+    if [ -z "$PY_INSTALLED" ]; then
+        for PY_VERS in "${PYTHON_VERSIONS[@]}"; do
```

Comment:
Can we make this either just install one specific version (instead of looping through different versions on failure either exit or retry) or is there a strong argument we can write clearly here in a comment that indicates why we are looping through different versions?

Reply:
Sure, I'll make it try to install just 3.10 if any of the 3.10-3.12 versions not already installed in the user's system. That would be more consistent with the pylint workflow testing 3.10